### DISCUSSION TITLE
How to define a column represented by a text expression

### DISCUSSION THREAD

**Discussion by nicolaerosia** [sqlalchemy/sqlalchemy#6625] [Usage Questions]
Hello,

In my application, users need to specify a text representing the expression which I use to get values representing a timestamp, example user input:
```
TIME_SLICE(INSERT_TIME, 30, 'MINUTE', 'END')
```
In this example the existing table column is `INSERT_TIME` but in order to get aggregated data, the user specifies that expression.

I would like to use his expression to specify a label for this column but I don't understand how, here's what I would like to end up with:
```sql
SELECT
TIME_SLICE(INSERT_TIME, 30, 'MINUTE', 'END') AS internal_ts,
SUM(entries) AS y
FROM
<a-table-specified-by-the-user)
WHERE
INSERT_TIME > <my-filter>
GROUP BY internal_ts
```

Creating a column on user input `TIME_SLICE(INSERT_TIME, 30, 'MINUTE', 'END') ` will escape this string and the SQL server will interpret it as an existing field in the column.

Using `.label()` on `text()` does not work either:
```python
ts_field_name = "INSERT_TIME"
ts_expr = "TIME_SLICE(INSERT_TIME, 30, 'MINUTE', 'END')"
ts_target_name = "internal_ts"
ts_expr_col = text(ts_expr).label(ts_target_name)

table = "my_table"

ts_col = Column(ts_field_name, DateTime())
stmt = select(ts_expr_col).select_from(text(table))
stmt = stmt.where(ts_col > min_ts)
stmt = stmt.group_by(ts_expr_col)
```

Error:
```
AttributeError: 'TextClause' object has no attribute 'label'
```

Any suggestions/ideas on how I should approach this (I can always add myself the string " AS internal_ts" but I'm looking for a sqlalchemy way :) )

**Comment by CaselIT** [comment]
Hi,

I would suggest using lowercase [column](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.column) (or [literal_column](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.literal_column)) and [table](https://docs.sqlalchemy.org/en/14/core/selectable.html#sqlalchemy.sql.expression.table) instead of text in your examples. Also as a replacement for `Column`

You may find useful to use [func](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.func) to represent the function call

**Reply by nicolaerosia** [reply]
@CaselIT  thank you for looking into this! Here's my code now:
```python
ts_field_name = "INSERT_TIME"
ts_expr = "TIME_SLICE(INSERT_TIME, 30, 'MINUTE', 'END')"
ts_target_name = "internal_ts"
ts_expr_col = literal_column(ts_expr).label(ts_target_name)

table_name = "my_table"

ts_col = column(ts_field_name, DateTime())
stmt = select(ts_expr_col).select_from(table(table_name))
stmt = stmt.where(ts_col > min_ts)
stmt = stmt.group_by(ts_expr_col)
```
Which generates:
```
'SELECT TIME_SLICE(INSERT_TIME, 30, \'MINUTE\', \'END\') AS internal_ts \nFROM my_table \nWHERE "INSERT_TIME" > \'2021-06-13 08:42:13.138307\' GROUP BY TIME_SLICE(INSERT_TIME, 30, \'MINUTE\', \'END\')'
```
It seems that the `group_by` on `ts_expr_col` does not use the label - is this correct?
Anyway - it looks much better now - I can just `group_by` on a `column` with that name.

I'm using this function to debug:
```python
def sqlalchemy_stmt_to_str2(stmt: Any) -> str:
    stmt_compiled = stmt.compile(compile_kwargs={"literal_binds": True})

    return str(stmt_compiled)
```

**Reply by CaselIT** [reply]
that may depend on the db used. To see what is the actual sql executed by sqlalchemy you can toggle the [`echo`](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=create_engine#sqlalchemy.create_engine.params.echo) flag on the engine

**Reply by nicolaerosia** [reply]
ahhh, ok, I will definitely try that. Is there a reliable way to know what would get executed without echo and actually executing it ?:)

**Reply by CaselIT** [reply]
You can pass an instance of the dialect to the compile function, like so:
```py
from sqlalchemy.dialect.mssql import dialect
print (str(stmt.compile(dialect=dialect(), compile_kwargs={...}))
```
